File: bindings/go/configure/configure.go
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://github.com/openpitkit and the OWNERS file for details.

// Package configure provides runtime policy-settings updates bound to an engine.
package configure

import (
	"fmt"

	"go.openpit.dev/openpit/internal/native"
	"go.openpit.dev/openpit/param"
	"go.openpit.dev/openpit/pkg/optional"
	"go.openpit.dev/openpit/pkg/ptr"
	"go.openpit.dev/openpit/pretrade/policies"
)

// Configurator updates the runtime settings of built-in policies registered on
// an engine. Obtain it from an engine's Configure accessor. It carries no
// state of its own: every call forwards to the engine it was created from,
// and it is valid for as long as that engine is.
type Configurator struct {
	engine native.Engine
}

// NewFromHandle wraps a native engine handle into a Configurator accessor.
func NewFromHandle(engine native.Engine) Configurator {
	return Configurator{engine: engine}
}

//------------------------------------------------------------------------------
// Error

// ErrorKind classifies an Error.
type ErrorKind uint32

const (
	// ErrorKindUnknown means no configurable policy carries the requested name.
	ErrorKindUnknown ErrorKind = ErrorKind(native.ConfigureErrorKindUnknown)
	// ErrorKindTypeMismatch means the policy name matched a policy of a
	// different type than the configure call targets.
	ErrorKindTypeMismatch ErrorKind = ErrorKind(native.ConfigureErrorKindTypeMismatch)
	// ErrorKindValidation means the supplied configuration values failed
	// validation.
	ErrorKindValidation ErrorKind = ErrorKind(native.ConfigureErrorKindValidation)
)

// Error is returned when a runtime configure call fails.
//
// Kind classifies the failure; Message provides a human-readable description.
type Error struct {
	// Message is the human-readable error description.
	Message string
	// Kind classifies the failure.
	Kind ErrorKind
}

// Error implements the error interface.
func (e *Error) Error() string {
	return e.Message
}

func newErrorFromHandle(handle native.ConfigureError) *Error {
	msg := native.ConfigureErrorGetMessage(handle)
	kind := native.ConfigureErrorGetKind(handle)
	native.DestroyConfigureError(handle)
	return &Error{
		Message: msg,
		Kind:    ErrorKind(kind),
	}
}

//------------------------------------------------------------------------------
// RateLimit

// RateLimit updates the runtime settings of the named rate-limit policy.
//
// broker, assets, accounts, and accountAssets mirror the axis types accepted by
// [policies.RateLimitReadyBuilder]. A non-nil axis replaces that axis
// wholesale; barriers can be added and removed at runtime. A barrier key that
// survives the replacement keeps its live counter (no reset). An empty non-nil
// slice clears the axis, subject to the policy's at-least-one-barrier rule.
// Nil axes and nil broker are left unchanged.
//
// Returns a *ConfigureError on a domain error (kind TypeMismatch when the name
// belongs to a different policy type, Validation when values are invalid).
func (c Configurator) RateLimit(
	name string,
	broker *policies.RateLimitBrokerBarrier,
	assets []policies.RateLimitAssetBarrier,
	accounts []policies.RateLimitAccountBarrier,
	accountAssets []policies.RateLimitAccountAssetBarrier,
) error {
	var nativeBroker *native.PretradePoliciesRateLimitBrokerBarrier
	if broker != nil {
		windowNanos := uint64(broker.Limit.Window) //nolint:gosec // negative duration becomes large uint64; core rejects it
		b := native.NewPretradePoliciesRateLimitBrokerBarrier(broker.Limit.MaxOrders, windowNanos)
		nativeBroker = ptr.New(b)
	}

	var nativeAssets []native.PretradePoliciesRateLimitAssetBarrier
	if assets != nil {
		nativeAssets = make([]native.PretradePoliciesRateLimitAssetBarrier, 0, len(assets))
		for _, a := range assets {
			windowNanos := uint64(a.Limit.Window) //nolint:gosec // negative duration becomes large uint64; core rejects it
			nativeAssets = append(nativeAssets, native.NewPretradePoliciesRateLimitAssetBarrier(
				a.Limit.MaxOrders,
				windowNanos,
				a.SettlementAsset.Handle(),
			))
		}
	}

	var nativeAccounts []native.PretradePoliciesRateLimitAccountBarrier
	if accounts != nil {
		nativeAccounts = make([]native.PretradePoliciesRateLimitAccountBarrier, 0, len(accounts))
		for _, a := range accounts {
			windowNanos := uint64(a.Limit.Window) //nolint:gosec // negative duration becomes large uint64; core rejects it
			nativeAccounts = append(nativeAccounts, native.NewPretradePoliciesRateLimitAccountBarrier(
				a.AccountID.Handle(),
				a.Limit.MaxOrders,
				windowNanos,
			))
		}
	}

	var nativeAccountAssets []native.PretradePoliciesRateLimitAccountAssetBarrier
	if accountAssets != nil {
		nativeAccountAssets = make([]native.PretradePoliciesRateLimitAccountAssetBarrier, 0, len(accountAssets))
		for _, a := range accountAssets {
			windowNanos := uint64(a.Limit.Window) //nolint:gosec // negative duration becomes large uint64; core rejects it
			nativeAccountAssets = append(
				nativeAccountAssets,
				native.NewPretradePoliciesRateLimitAccountAssetBarrier(
					a.AccountID.Handle(),
					a.Limit.MaxOrders,
					windowNanos,
					a.SettlementAsset.Handle(),
				),
			)
		}
	}

	configErr := native.EngineConfigureRateLimit(
		c.engine,
		name,
		nativeBroker,
		nativeAssets,
		nativeAccounts,
		nativeAccountAssets,
	)
	if configErr != nil {
		return newErrorFromHandle(configErr)
	}
	return nil
}

//------------------------------------------------------------------------------
// PnlBoundsKillSwitch

// PnlBoundsKillSwitch updates the runtime settings of the named P&L bounds
// kill-switch policy.
//
// brokerBarriers mirrors the broker axis accepted by
// [policies.PnlBoundsKillswitchReadyBuilder]. accountBarriers updates bounds
// without replacing the live P&L accumulated for each account and settlement
// asset. An axis passed as nil is left unchanged; an empty non-nil slice
// replaces the axis with an empty set (subject to the policy's
// at-least-one-barrier rule).
//
// Returns a *Error on a domain error.
func (c Configurator) PnlBoundsKillSwitch(
	name string,
	brokerBarriers []policies.PnlBoundsBrokerBarrier,
	accountBarriers []policies.PnlBoundsAccountAssetBarrierUpdate,
) error {
	var nativeBrokers []native.PretradePoliciesPnlBoundsBarrier
	if brokerBarriers != nil {
		nativeBrokers = make([]native.PretradePoliciesPnlBoundsBarrier, 0, len(brokerBarriers))
		for _, b := range brokerBarriers {
			nativeBrokers = append(nativeBrokers, native.NewPretradePoliciesPnlBoundsBarrier(
				b.SettlementAsset.Handle(),
				pnlOptionalToNative(b.LowerBound),
				pnlOptionalToNative(b.UpperBound),
			))
		}
	}

	var nativeAccounts []native.PretradePoliciesPnlBoundsAccountBarrierUpdate
	if accountBarriers != nil {
		nativeAccounts = make(
			[]native.PretradePoliciesPnlBoundsAccountBarrierUpdate,
			0,
			len(accountBarriers),
		)
		for _, a := range accountBarriers {
			nativeAccounts = append(
				nativeAccounts,
				native.NewPretradePoliciesPnlBoundsAccountBarrierUpdate(
					a.AccountID.Handle(),
					a.Barrier.SettlementAsset.Handle(),
					pnlOptionalToNative(a.Barrier.LowerBound),
					pnlOptionalToNative(a.Barrier.UpperBound),
				),
			)
		}
	}

	configErr := native.EngineConfigurePnlBoundsKillSwitch(
		c.engine,
		name,
		nativeBrokers,
		nativeAccounts,
	)
	if configErr != nil {
		return newErrorFromHandle(configErr)
	}
	return nil
}

//------------------------------------------------------------------------------
// SetAccountPnl

// SetAccountPnl force-sets the live accumulated P&L for one
// (account, settlementAsset) entry of the named P&L bounds kill-switch policy.
//
// This is an absolute assignment (upsert): the entry is created if it does not
// exist yet, exactly as a construction-time seed would. It is distinct from
// [Configurator.PnlBoundsKillSwitch], which retunes bounds and never touches
// accumulated P&L. The new value is evaluated against the live bounds on the
// next check; forcing the accumulator past a bound trips the kill switch and
// latches an engine-level account block that this call never clears.
//
// Returns a *Error on a domain error (kind TypeMismatch when the name belongs
// to a different policy type, Unknown when no policy carries the name).
func (c Configurator) SetAccountPnl(
	name string,
	account param.AccountID,
	settlementAsset param.Asset,
	pnl param.Pnl,
) error {
	configErr := native.EngineSetAccountPnl(
		c.engine,
		name,
		account.Handle(),
		settlementAsset.Handle(),
		pnl.Handle(),
	)
	if configErr != nil {
		return newErrorFromHandle(configErr)
	}
	return nil
}

//------------------------------------------------------------------------------
// OrderSizeLimit

// OrderSizeLimit updates the runtime settings of the named order-size-limit
// policy.
//
// broker, assets, and accountAssets mirror the axis types accepted by
// [policies.OrderSizeLimitReadyBuilder]. An axis passed as nil is left
// unchanged; an empty non-nil slice replaces the axis with an empty set
// (subject to the policy's at-least-one-barrier rule). A nil broker leaves
// the broker barrier unchanged.
//
// Returns a *Error on a domain error.
func (c Configurator) OrderSizeLimit(
	name string,
	broker *policies.OrderSizeBrokerBarrier,
	assets []policies.OrderSizeAssetBarrier,
	accountAssets []policies.OrderSizeAccountAssetBarrier,
) error {
	var nativeBroker *native.PretradePoliciesOrderSizeBrokerBarrier
	if broker != nil {
		b := native.NewPretradePoliciesOrderSizeBrokerBarrier(
			native.NewPretradePoliciesOrderSizeLimit(
				broker.Limit.MaxQuantity.Handle(),
				broker.Limit.MaxNotional.Handle(),
			),
		)
		nativeBroker = ptr.New(b)
	}

	var nativeAssets []native.PretradePoliciesOrderSizeAssetBarrier
	if assets != nil {
		nativeAssets = make([]native.PretradePoliciesOrderSizeAssetBarrier, 0, len(assets))
		for _, a := range assets {
			nativeAssets = append(nativeAssets, native.NewPretradePoliciesOrderSizeAssetBarrier(
				native.NewPretradePoliciesOrderSizeLimit(
					a.Limit.MaxQuantity.Handle(),
					a.Limit.MaxNotional.Handle(),
				),
				a.SettlementAsset.Handle(),
			))
		}
	}

	var nativeAccountAssets []native.PretradePoliciesOrderSizeAccountAssetBarrier
	if accountAssets != nil {
		nativeAccountAssets = make([]native.PretradePoliciesOrderSizeAccountAssetBarrier, 0, len(accountAssets))
		for _, a := range accountAssets {
			nativeAccountAssets = append(
				nativeAccountAssets,
				native.NewPretradePoliciesOrderSizeAccountAssetBarrier(
					native.NewPretradePoliciesOrderSizeLimit(
						a.Limit.MaxQuantity.Handle(),
						a.Limit.MaxNotional.Handle(),
					),
					a.AccountID.Handle(),
					a.SettlementAsset.Handle(),
				),
			)
		}
	}

	configErr := native.EngineConfigureOrderSizeLimit(
		c.engine,
		name,
		nativeBroker,
		nativeAssets,
		nativeAccountAssets,
	)
	if configErr != nil {
		return newErrorFromHandle(configErr)
	}
	return nil
}

//------------------------------------------------------------------------------
// SpotFunds

// SpotFunds updates the runtime settings of the named spot-funds policy.
//
// globalSlippageBps and pricingSource are optional: pass None to leave them
// unchanged. Each override entry is applied individually (insert-or-clear):
// an entry whose SlippageBps is None clears any override at its target. A nil
// overrides slice leaves the cascade untouched; entries never replace the
// whole table.
//
// The overrides slice and [policies.SpotFundsOverrideEntry] type mirror those
// accepted by [policies.SpotFundsReadyBuilder.Overrides].
//
// Returns a *Error on a domain error.
func (c Configurator) SpotFunds(
	name string,
	globalSlippageBps optional.Option[uint16],
	pricingSource optional.Option[policies.SpotFundsPricingSource],
	overrides []policies.SpotFundsOverrideEntry,
) error {
	var slippagePtr *uint16
	if v, has := globalSlippageBps.Get(); has {
		slippagePtr = ptr.New(v)
	}

	var pricingSourcePtr *uint8
	if v, has := pricingSource.Get(); has {
		u := uint8(v)
		pricingSourcePtr = &u
	}

	var nativeOverrides []native.PretradePoliciesSpotFundsOverride
	if overrides != nil {
		nativeOverrides = make([]native.PretradePoliciesSpotFundsOverride, 0, len(overrides))
		for i, e := range overrides {
			var slippageBpsPtr *uint16
			if v, has := e.Override.SlippageBps.Get(); has {
				slippageBpsPtr = ptr.New(v)
			}
			override, err := policies.NewNativeSpotFundsOverride(e.Target, slippageBpsPtr)
			if err != nil {
				return &Error{
					Kind:    ErrorKindValidation,
					Message: fmt.Sprintf("configure: spot funds override %d: %v", i, err),
				}
			}
			nativeOverrides = append(nativeOverrides, override)
		}
	}

	configErr := native.EngineConfigureSpotFunds(
		c.engine,
		name,
		slippagePtr,
		pricingSourcePtr,
		nativeOverrides,
	)
	if configErr != nil {
		return newErrorFromHandle(configErr)
	}
	return nil
}

//------------------------------------------------------------------------------
// Helpers

func pnlOptionalToNative(value optional.Option[param.Pnl]) native.ParamPnlOptional {
	if v, has := value.Get(); has {
		return native.NewParamPnlOptional(v.Handle())
	}
	return native.ParamPnlOptional{}
}

File: crates/openpit/src/core/configure.rs
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://github.com/openpitkit and the OWNERS file for details.

//! Runtime reconfiguration for built-in pre-trade policies.
//!
//! At build time the engine captures a clone of each supported built-in
//! policy's settings cell. The registry stores those cells as concrete enum
//! variants keyed by policy name. A [`ConfigCell`](crate::storage::ConfigCell)
//! clone shares the running policy's value, so updates are observed on the next
//! hot-path read without adding synchronization to order checks.
//!
//! Runtime reconfiguration of custom policies is not supported in this
//! release.

use std::cell::Cell;
use std::collections::HashMap;
use std::fmt::{Display, Formatter};

use super::sync_mode::SyncMode;
use crate::param::{AccountId, Asset, Pnl};
use crate::pretrade::policies::{
    OrderSizeLimitSettings, PnlBoundsKillSwitchSettings, RateLimitSettings, RealizedPnlStorage,
    SpotFundsSettings,
};
use crate::storage::{ConfigCell, LockingPolicyFactory};

// ─── ConfigEntry ────────────────────────────────────────────────────────────

/// Settings cells supported by the built-in runtime configurator.
pub(crate) enum ConfigEntry<Factory: LockingPolicyFactory> {
    /// Rate-limit policy settings.
    RateLimit(Factory::Config<RateLimitSettings>),
    /// P&L bounds kill-switch policy handles.
    ///
    /// Carries both the settings cell (bounds retune) and the shared realized
    /// P&L ledger (force-set of accumulated P&L), so the configurator can reach
    /// either without an extra registry lookup.
    PnlBoundsKillSwitch {
        /// Settings cell shared with the running policy.
        settings: Factory::Config<PnlBoundsKillSwitchSettings>,
        /// Live accumulated P&L ledger shared with the running policy.
        realized: RealizedPnlStorage<Factory>,
    },
    /// Spot-funds policy settings.
    SpotFunds(Factory::Config<SpotFundsSettings>),
    /// Order-size-limit policy settings.
    OrderSizeLimit(Factory::Config<OrderSizeLimitSettings>),
}

impl<Factory: LockingPolicyFactory> ConfigEntry<Factory> {
    fn settings_type_name(&self) -> &'static str {
        match self {
            Self::RateLimit(_) => std::any::type_name::<Factory::Config<RateLimitSettings>>(),
            Self::PnlBoundsKillSwitch { .. } => {
                std::any::type_name::<Factory::Config<PnlBoundsKillSwitchSettings>>()
            }
            Self::SpotFunds(_) => std::any::type_name::<Factory::Config<SpotFundsSettings>>(),
            Self::OrderSizeLimit(_) => {
                std::any::type_name::<Factory::Config<OrderSizeLimitSettings>>()
            }
        }
    }
}

// ─── ConfigureError ──────────────────────────────────────────────────────────

/// Error returned by [`Configurator`] when a runtime reconfiguration fails.
///
/// Every variant leaves the live settings unchanged: an unknown or mismatched
/// policy is never touched, and a rejected update is rolled back before
/// publication (the [`ConfigCell`](crate::storage::ConfigCell) update is
/// transactional).
#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub enum ConfigureError {
    /// No registered policy carries the requested name.
    UnknownPolicy {
        /// Requested policy name.
        name: String,
    },
    /// A policy is registered under `name`, but its settings type differs from
    /// the one the called method targets.
    PolicyTypeMismatch {
        /// Requested policy name.
        name: String,
        /// Settings cell type the method expected.
        expected: &'static str,
        /// Settings cell type actually registered.
        found: &'static str,
    },
    /// The update closure rejected the new value; the prior value still
    /// applies.
    Validation {
        /// Requested policy name.
        name: String,
        /// Rendered error returned by the closure.
        message: String,
    },
    /// A configuration call was issued from within another configuration
    /// callback on the same thread.
    ///
    /// Configuration is non-reentrant: a [`Configurator`] method runs its
    /// update closure while it owns the settings cell's writer lock, and that
    /// lock is not reentrant. Re-entering configuration from inside such a
    /// closure - whether for the same policy or a different one - would
    /// deadlock, so it is rejected before any lock is taken. Configuration
    /// from other threads is unaffected and still serializes. The live
    /// settings are left unchanged.
    NestedConfiguration,
}

impl Display for ConfigureError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::UnknownPolicy { name } => {
                write!(formatter, "no configurable policy named {name}")
            }
            Self::PolicyTypeMismatch {
                name,
                expected,
                found,
            } => write!(
                formatter,
                "policy {name} has settings type {found}, not {expected}"
            ),
            Self::Validation { name, message } => {
                write!(formatter, "policy {name} rejected the update: {message}")
            }
            Self::NestedConfiguration => write!(
                formatter,
                "configuration is not reentrant: cannot configure from within \
                 another configuration callback on the same thread"
            ),
        }
    }
}

impl std::error::Error for ConfigureError {}

// ─── ConfigRegistry ──────────────────────────────────────────────────────────

/// Per-engine map from policy name to its built-in settings cell.
pub(crate) struct ConfigRegistry<Factory: LockingPolicyFactory> {
    entries: HashMap<String, ConfigEntry<Factory>>,
}

impl<Factory: LockingPolicyFactory> ConfigRegistry<Factory> {
    /// Builds a registry from the builder's collected `(name, cell)` pairs.
    pub(crate) fn from_entries(entries: HashMap<String, ConfigEntry<Factory>>) -> Self {
        Self { entries }
    }

    fn entry(&self, name: &str) -> Result<&ConfigEntry<Factory>, ConfigureError> {
        self.entries
            .get(name)
            .ok_or_else(|| ConfigureError::UnknownPolicy {
                name: name.to_owned(),
            })
    }

    fn type_mismatch<Settings: Clone + 'static>(
        name: &str,
        entry: &ConfigEntry<Factory>,
    ) -> ConfigureError {
        ConfigureError::PolicyTypeMismatch {
            name: name.to_owned(),
            expected: std::any::type_name::<Factory::Config<Settings>>(),
            found: entry.settings_type_name(),
        }
    }

    fn validation<Error: std::fmt::Display>(name: &str, error: Error) -> ConfigureError {
        ConfigureError::Validation {
            name: name.to_owned(),
            message: error.to_string(),
        }
    }
}

// ─── Re-entrancy guard ───────────────────────────────────────────────────────

thread_local! {
    // Set while a configuration call is in progress on the current thread.
    // Guards every `Configurator` entry against re-entrant configuration,
    // which would deadlock on a settings cell's non-reentrant writer lock
    // (see `ConfigureError::NestedConfiguration`).
    static CONFIGURING: Cell<bool> = const { Cell::new(false) };
}

/// RAII marker that a configuration is in progress on the current thread.
///
/// [`Self::enter`] fails with [`ConfigureError::NestedConfiguration`] when a
/// configuration is already active on this thread, so a nested call returns
/// before it can take any cell writer lock. The flag is cleared on drop -
/// including on an early `?` return or a panic unwinding through the closure -
/// so a single thread can configure again once the outer call completes.
struct ConfiguringGuard;

impl ConfiguringGuard {
    fn enter() -> Result<Self, ConfigureError> {
        CONFIGURING.with(|configuring| {
            if configuring.get() {
                return Err(ConfigureError::NestedConfiguration);
            }
            configuring.set(true);
            Ok(Self)
        })
    }
}

impl Drop for ConfiguringGuard {
    fn drop(&mut self) {
        CONFIGURING.with(|configuring| configuring.set(false));
    }
}

// ─── Configurator ────────────────────────────────────────────────────────────

/// Storage locking factory backing a [`Configurator`] of a given [`SyncMode`].
type RegistryFactory<Sync> = <Sync as SyncMode>::StorageLockingPolicyFactory;

/// Engine handle for retuning built-in policies at runtime.
///
/// Obtained from [`Engine::configure`](crate::Engine::configure). Each method
/// targets one supported built-in policy settings type. Updates published here
/// are observed by the running engine on its next hot-path read because the
/// registry shares each policy's settings cell rather than holding a copy.
/// Custom-policy runtime reconfiguration is planned for a later release.
///
/// # Threading
///
/// The handle's auto-traits derive from the engine's [`SyncMode`], matching the
/// engine handle:
///
/// - [`FullSync`](crate::FullSync): `Send + Sync`.
/// - [`AccountSync`](crate::AccountSync): `Send + !Sync`.
/// - [`LocalSync`](crate::LocalSync): `!Send + !Sync`.
pub struct Configurator<Sync: SyncMode> {
    registry: <<Sync as SyncMode>::StorageLockingPolicyFactory as LockingPolicyFactory>::Shared<
        ConfigRegistry<<Sync as SyncMode>::StorageLockingPolicyFactory>,
    >,
}

impl<Sync: SyncMode> Clone for Configurator<Sync> {
    fn clone(&self) -> Self {
        Self {
            registry: self.registry.clone(),
        }
    }
}

impl<Sync: SyncMode> Configurator<Sync> {
    /// Wraps the engine's shared registry in a configurator handle.
    pub(crate) fn from_inner(
        registry: <<Sync as SyncMode>::StorageLockingPolicyFactory as LockingPolicyFactory>::Shared<
            ConfigRegistry<<Sync as SyncMode>::StorageLockingPolicyFactory>,
        >,
    ) -> Self {
        Self { registry }
    }

    /// Retunes the [`RateLimitPolicy`](crate::pretrade::policies::RateLimitPolicy)
    /// registered under `name`.
    ///
    /// # Errors
    ///
    /// Returns [`ConfigureError::UnknownPolicy`] for an unknown name,
    /// [`ConfigureError::PolicyTypeMismatch`] when the name belongs to another
    /// built-in policy type, [`ConfigureError::Validation`] when `f` rejects
    /// the update, or [`ConfigureError::NestedConfiguration`] when called from
    /// within another configuration callback on the same thread.
    pub fn rate_limit<Error: std::fmt::Display>(
        &self,
        name: &str,
        f: impl FnOnce(&mut RateLimitSettings) -> Result<(), Error>,
    ) -> Result<(), ConfigureError> {
        let _guard = ConfiguringGuard::enter()?;
        match self.registry.entry(name)? {
            ConfigEntry::RateLimit(cell) => cell
                .update(f)
                .map_err(|error| ConfigRegistry::<RegistryFactory<Sync>>::validation(name, error)),
            entry => Err(ConfigRegistry::<_>::type_mismatch::<RateLimitSettings>(
                name, entry,
            )),
        }
    }

    /// Retunes the
    /// [`PnlBoundsKillSwitchPolicy`](crate::pretrade::policies::PnlBoundsKillSwitchPolicy)
    /// registered under `name`.
    ///
    /// # Errors
    ///
    /// Returns the same error variants as [`Self::rate_limit`].
    pub fn pnl_bounds_killswitch<Error: std::fmt::Display>(
        &self,
        name: &str,
        f: impl FnOnce(&mut PnlBoundsKillSwitchSettings) -> Result<(), Error>,
    ) -> Result<(), ConfigureError> {
        let _guard = ConfiguringGuard::enter()?;
        match self.registry.entry(name)? {
            ConfigEntry::PnlBoundsKillSwitch { settings, .. } => settings
                .update(f)
                .map_err(|error| ConfigRegistry::<RegistryFactory<Sync>>::validation(name, error)),
            entry => Err(ConfigRegistry::<_>::type_mismatch::<
                PnlBoundsKillSwitchSettings,
            >(name, entry)),
        }
    }

    /// Force-sets the live accumulated P&L for a `(account, settlement_asset)`
    /// entry of the
    /// [`PnlBoundsKillSwitchPolicy`](crate::pretrade::policies::PnlBoundsKillSwitchPolicy)
    /// registered under `name`.
    ///
    /// This is an absolute assignment (upsert): the entry is created if it does
    /// not exist yet, exactly as a construction-time seed would. It is distinct
    /// from [`Self::pnl_bounds_killswitch`], which retunes the bounds and never
    /// touches accumulated P&L. The new value is evaluated against the live
    /// bounds on the next hot-path read.
    ///
    /// Unlike the settings retune, this does not take the configuration
    /// re-entrancy guard and does not touch the settings writer mutex: it writes
    /// only the realized-P&L storage (its own lock) and runs no user closure, so
    /// it can neither deadlock nor re-enter configuration.
    ///
    /// # Errors
    ///
    /// Returns [`ConfigureError::UnknownPolicy`] for an unknown name or
    /// [`ConfigureError::PolicyTypeMismatch`] when the name belongs to another
    /// built-in policy type. The operation is otherwise infallible: an absolute
    /// assignment performs no validation and cannot overflow.
    pub fn set_account_pnl(
        &self,
        name: &str,
        account: AccountId,
        settlement_asset: Asset,
        pnl: Pnl,
    ) -> Result<(), ConfigureError> {
        match self.registry.entry(name)? {
            ConfigEntry::PnlBoundsKillSwitch { realized, .. } => {
                // Absolute force-set: upsert the entry, mirroring the
                // construction-time seed. This writes only the realized ledger
                // (its own lock) and runs no user closure, so it cannot
                // re-enter configuration or deadlock on the settings writer.
                realized.with_mut(
                    (account, settlement_asset),
                    || Pnl::ZERO,
                    |entry, _is_new| *entry = pnl,
                );
                Ok(())
            }
            entry => Err(ConfigRegistry::<_>::type_mismatch::<
                PnlBoundsKillSwitchSettings,
            >(name, entry)),
        }
    }

    /// Retunes the [`SpotFundsPolicy`](crate::pretrade::policies::SpotFundsPolicy)
    /// registered under `name`.
    ///
    /// # Errors
    ///
    /// Returns the same error variants as [`Self::rate_limit`].
    pub fn spot_funds<Error: std::fmt::Display>(
        &self,
        name: &str,
        f: impl FnOnce(&mut SpotFundsSettings) -> Result<(), Error>,
    ) -> Result<(), ConfigureError> {
        let _guard = ConfiguringGuard::enter()?;
        match self.registry.entry(name)? {
            ConfigEntry::SpotFunds(cell) => cell
                .update(f)
                .map_err(|error| ConfigRegistry::<RegistryFactory<Sync>>::validation(name, error)),
            entry => Err(ConfigRegistry::<_>::type_mismatch::<SpotFundsSettings>(
                name, entry,
            )),
        }
    }

    /// Retunes the
    /// [`OrderSizeLimitPolicy`](crate::pretrade::policies::OrderSizeLimitPolicy)
    /// registered under `name`.
    ///
    /// # Errors
    ///
    /// Returns the same error variants as [`Self::rate_limit`].
    pub fn order_size_limit<Error: std::fmt::Display>(
        &self,
        name: &str,
        f: impl FnOnce(&mut OrderSizeLimitSettings) -> Result<(), Error>,
    ) -> Result<(), ConfigureError> {
        let _guard = ConfiguringGuard::enter()?;
        match self.registry.entry(name)? {
            ConfigEntry::OrderSizeLimit(cell) => cell
                .update(f)
                .map_err(|error| ConfigRegistry::<RegistryFactory<Sync>>::validation(name, error)),
            entry => Err(ConfigRegistry::<_>::type_mismatch::<OrderSizeLimitSettings>(name, entry)),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::cell::RefCell;
    use std::time::Duration;

    use crate::param::{AccountId, Asset, Quantity, Side, TradeAmount};
    use crate::pretrade::policies::{
        RateLimit, RateLimitBrokerBarrier, RateLimitPolicy, RateLimitPolicyError, RateLimitSettings,
    };
    use crate::storage::FullLocking;
    use crate::{Engine, FullSyncEngine, Instrument, OrderOperation};

    use super::ConfigureError;

    fn broker_barrier(max_orders: usize) -> RateLimitBrokerBarrier {
        RateLimitBrokerBarrier {
            limit: RateLimit {
                max_orders,
                // Wide window so every order in the test shares one window.
                window: Duration::from_secs(60),
            },
        }
    }

    fn build_engine(max_orders: usize) -> FullSyncEngine<OrderOperation> {
        let builder = Engine::builder::<OrderOperation, (), ()>().full_sync();
        let settings = RateLimitSettings::new(Some(broker_barrier(max_orders)), [], [], [])
            .expect("broker barrier is a valid configuration");
        let policy = RateLimitPolicy::<FullLocking>::new(settings, builder.storage_builder());
        builder
            .pre_trade(policy)
            .build()
            .expect("engine must build")
    }

    fn order(account: u64) -> OrderOperation {
        OrderOperation {
            instrument: Instrument::new(
                Asset::new("AAPL").expect("asset code must be valid"),
                Asset::new("USD").expect("asset code must be valid"),
            ),
            account_id: AccountId::from_u64(account),
            side: Side::Buy,
            trade_amount: TradeAmount::Quantity(
                Quantity::from_str("1").expect("quantity literal must be valid"),
            ),
            price: None,
        }
    }

    // A configuration call issued from within another configuration callback on
    // the same thread must be rejected with `NestedConfiguration` rather than
    // deadlocking on the settings cell's non-reentrant writer lock. The fact
    // that this test terminates is itself the no-hang proof.
    #[test]
    fn nested_same_thread_configuration_is_rejected_without_deadlock() {
        let engine = build_engine(2);
        let name = RateLimitPolicy::<FullLocking>::NAME;

        // Exhaust the broker limit of 2: the two orders pass (count 1..=2).
        for account in 0..2 {
            engine
                .execute_pre_trade(order(account))
                .expect("order within the limit must pass");
        }

        // The closure captures the engine by `&` (configure takes `&self`).
        // It first widens its private copy to 9, then re-enters configuration
        // for the same policy. The nested call must fail fast; the captured
        // error is asserted below. Returning the nested error rolls back the
        // outer transaction, so neither the widening to 9 nor the nested 100
        // is ever published.
        let nested = RefCell::new(None);
        let outer = engine
            .configure()
            .rate_limit::<ConfigureError>(name, |settings| {
                settings
                    .set_broker(Some(broker_barrier(9)))
                    .expect("widening to 9 is a valid private-copy edit");
                let inner = engine
                    .configure()
                    .rate_limit::<RateLimitPolicyError>(name, |settings| {
                        settings.set_broker(Some(broker_barrier(100)))
                    });
                let error = inner.expect_err("nested configuration must be rejected");
                *nested.borrow_mut() = Some(error.clone());
                Err(error)
            });

        // The nested call returned the new variant directly.
        assert_eq!(
            nested.into_inner(),
            Some(ConfigureError::NestedConfiguration)
        );
        // The outer call rolled back, surfacing the nested error through its
        // own validation channel.
        assert_eq!(
            outer,
            Err(ConfigureError::Validation {
                name: name.to_owned(),
                message: ConfigureError::NestedConfiguration.to_string(),
            })
        );

        // Live settings are unchanged: the limit is still 2, so a third order
        // (count 3) is rejected. A published widening would have admitted it.
        let rejects = engine
            .execute_pre_trade(order(2))
            .err()
            .expect("limit must still be 2 after the rejected retune");
        assert_eq!(rejects[0].reason, "rate limit exceeded: broker barrier");
    }
}
